voice/src/main/java/discord4j/voice/AudioProvider.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * This file is part of Discord4J.
+ *
+ * Discord4J is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Discord4J is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Discord4J.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package discord4j.voice;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Used to send audio.
+ * <p>
+ * The provider uses a shared buffer. Keep this in mind when implementing.
+ *
+ * @see #provide()
+ */
+public abstract class AudioProvider {
```

Comment:
I think this should actually be an interface. With the ByteBuffer utilities a part of a class named something like `AbstractAudioProvider`

Reply:
Is there any benefit to that?

Reply:
Interfaces have the potential to be a FunctionalInterface, and it allows for multiple inheritance.

Reply:
There is no way it could be a FunctionalInterface. The signature of provide() is dependent on there being a shared buffer.

Reply:
Why not have provide() return a ByteBuf which may or may not be a shared instance?

Reply:
I think it should be as easy as possible to do the most efficient thing which is the shared buffer. Returning a buffer encourages non-shared more than anything. 

I also can't think of a useful AudioProvider implementation which could actually just be a lambda. In v2, where IAudioProvider is a functional interface, I only ever saw classes implementing it. (See the v2 LavaPlayer demo).